src/backend/distributed/commands/table.c

```diff
@@ -593,59 +620,242 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
+ * contains a citus table with given type.
+ */
+static bool
+RelationIdListContainsCitusTableType(List *relationIdList, CitusTableType citusTableType)
+{
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, relationIdList)
 	{
-		/* left relation is not a postgres local table, */
-		return;
+		if (IsCitusTableType(relationId, citusTableType))
+		{
+			return true;
+		}
 	}
 
-	List *alterTableFKeyConstraints =
-		GetAlterTableStmtFKeyConstraintList(alterTableStatement);
-	Constraint *constraint = NULL;
-	foreach_ptr(constraint, alterTableFKeyConstraints)
+	return false;
+}
+
+
+/*
+ * RelationIdListContainsPostgresTable returns true if given relationIdList
+ * contains a postgres table.
+ */
+static bool
+RelationIdListContainsPostgresTable(List *relationIdList)
+{
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, relationIdList)
 	{
-		Oid rightRelationId = RangeVarGetRelid(constraint->pktable, lockmode,
-											   alterTableStatement->missing_ok);
-		if (IsCitusTableType(rightRelationId, CITUS_LOCAL_TABLE))
+		if (OidIsValid(relationId) && !IsCitusTable(relationId))
+		{
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * ConvertPostgresLocalTablesToCitusLocalTables converts each postgres table
+ * involved in foreign keys to be defined by given alter table command and the
+ * other tables connected to them via a foreign key graph to citus local tables.
+ */
+static void
+ConvertPostgresLocalTablesToCitusLocalTables(AlterTableStmt *alterTableStatement)
+{
+	List *rightRelationRangeVarList =
+		GetAlterTableAddFKeyRightRelationRangeVarList(alterTableStatement);
+	RangeVar *leftRelationRangeVar = alterTableStatement->relation;
+	List *relationRangeVarList = lappend(rightRelationRangeVarList, leftRelationRangeVar);
+
+	/*
+	 * To prevent deadlocks, sort the list before converting each postgres local
+	 * table to a citus local table.
+	 */
+	relationRangeVarList = SortList(relationRangeVarList, CompareRangeVarsByOid);
+
+	/*
+	 * Here we should operate on RangeVar objects since relations oid's would
+	 * change in below loop due to CreateCitusLocalTable.
+	 */
+	RangeVar *relationRangeVar;
+	foreach_ptr(relationRangeVar, relationRangeVarList)
+	{
+		List *commandList = alterTableStatement->cmds;
+		LOCKMODE lockMode = AlterTableGetLockLevel(commandList);
+		bool missingOk = alterTableStatement->missing_ok;
+		Oid relationId = RangeVarGetRelid(relationRangeVar, lockMode, missingOk);
+		if (!OidIsValid(relationId))
+		{
+			/*
+			 * As we are in preprocess, missingOk might be true and relation
+			 * might not exist.
+			 */
+			continue;
+		}
+		else if (IsCitusTable(relationId))
+		{
+			/*
+			 * relationRangeVarList has also reference and citus local tables
+			 * involved in this ADD FOREIGN KEY command. Moreover, even if
+			 * relationId was belonging to a postgres  local table initially,
+			 * we might had already converted it to a citus local table by cascading.
+			 */
+			continue;
+		}
+
+		/*
```

Comment:
lets mention that we have this try/catch only for better error messages

Reply:
cd5647b56a86dc2a592bbaa3df92e5f73536d431